yt/frontends/amrvac/data_structures.py

```diff
@@ -242,24 +243,106 @@ def _parse_parameter_file(self):
+
+        # 4. calculations for magnetic unit and time
+        time_unit = getattr(self, 'time_unit', length_unit / velocity_unit)  # if time given use it, else calculate
+        magnetic_unit = (np.sqrt(4 * np.pi * pressure_unit)).to('gauss')
+
+        setdefaultattr(self, 'mass_unit', mass_unit)
+        setdefaultattr(self, 'density_unit', density_unit)
+        setdefaultattr(self, 'numberdensity_unit', numberdensity_unit)
 
-        setdefaultattr(self, "length_unit", self.quan(1, "cm"))
-        setdefaultattr(self, "mass_unit", self.quan(1, "g"))
-        setdefaultattr(self, "time_unit", self.quan(1, "s"))
+        setdefaultattr(self, 'length_unit', length_unit)
+        setdefaultattr(self, 'velocity_unit', velocity_unit)
+        setdefaultattr(self, 'time_unit', time_unit)
 
-        # derived units
-        setdefaultattr(self, "velocity_unit", self.quan(self.length_unit/self.time_unit))
-        setdefaultattr(self, "density_unit", self.quan(self.mass_unit/self.length_unit**3))
-        setdefaultattr(self, "numberdensity_unit", self.quan(self.length_unit**-3))
+        setdefaultattr(self, 'temperature_unit', temperature_unit)
+        setdefaultattr(self, 'pressure_unit', pressure_unit)
+        setdefaultattr(self, 'magnetic_unit', magnetic_unit)
 
-        # TODO: check this
-        setdefaultattr(self, "temperature_unit", self.quan(1, "K"))
-        setdefaultattr(self, "pressure_unit", self.quan(1, "dyn*cm**-2"))
-        setdefaultattr(self, "magnetic_unit", self.quan(1, "gauss"))
+
+    def _override_code_units(self):
+        self._check_override_consistency()
+        super(AMRVACDataset, self)._override_code_units()
+
+
+    def _check_override_consistency(self):
+        # frontend specific method
+        # YT supports overriding other normalisations, this method ensures consistency between
+        # supplied 'units_override' items and those used by AMRVAC.
+
+        # AMRVAC's normlisations/units have 3 degrees of freedom.
+        # Moreover, if temperature unit is specified then velocity unit will be calculated
+        # accordingly, and vice-versa.
+        # Currently we replicate this by allowing a finite set of combinations in units_override
+        if not self.units_override:
+            return
+        overrides = set(self.units_override)
+
+        # there are only three degrees of freedom, so explicitly check for this
+        if len(overrides) > 3:
+            raise ValueError('More than 3 degrees of freedom were specified '
+                             'in units_override ({} given)'.format(len(overrides)))
+        # temperature and velocity cannot both be specified
+        if 'temperature_unit' in overrides and 'velocity_unit' in overrides:
+            raise ValueError('Either temperature or velocity is allowed in units_override, not both.')
+
+        allowed_combinations = [{'numberdensity_unit', 'temperature_unit', 'length_unit'},
```

Comment:
nit: I would define at module level and give it an all-uppercase name like `ALLOWED_UNIT_COMBINATIONS` to make it clearer this is constant data.

Reply:
Done as suggested!